crates/pipeline_manager/src/main.rs

```diff
@@ -78,10 +80,63 @@ use db::{
     storage::Storage, AttachedConnector, AttachedConnectorId, ConnectorId, DBError, PipelineId,
     ProgramDescr, ProgramId, ProjectDB, Version,
 };
-use runner::{LocalRunner, Runner, RunnerError};
+use runner::{LocalRunner, Runner, RunnerError, STARTUP_TIMEOUT};
 
 use crate::auth::TenantId;
 
+/// Errors validating API endpoint parameters.
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+pub enum ApiError {
+    ProgramNotSpecified,
+    PipelineNotSpecified,
+    ConnectorNotSpecified,
+    // I don't think this can ever happen.
+    MissingUrlEncodedParam { param: &'static str },
+    InvalidUuidParam { value: String, error: String },
+    InvalidPipelineAction { action: String },
+}
+
+impl StdError for ApiError {}
+
+impl Display for ApiError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self {
+            Self::ProgramNotSpecified => {
+                f.write_str("Program not specified: set either '?id=' or '?name='")
```

Comment:
This one and the next two can be more precise. "Program not specified. Use ?id or ?name query strings in the URL."